Add support for ImageCapture.grabFrame

youennf opened the pull request:
#### ae6cd18d5eb77dd68e72c04b2b7c887567ba3cf8
<pre>
Add support for ImageCapture.grabFrame
<a href="https://rdar.apple.com/148425176">rdar://148425176</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290916">https://bugs.webkit.org/show_bug.cgi?id=290916</a>

Reviewed by Eric Carlson.

Add support of ImageCapture::grabFrame by doing the following:
- Introduce ImageCaptureVideoFrameObserver as a VideoFrameObserver on the ImageCapture track.
- Convert the VideoFrame into an ImageBitmap. We handle the rotation so that drawing the ImageBitmap is similar to rendering the track via a media element.
- The conversion requires conversion to RGB which is done in GPUProcess. For that purpose, we introduce a new MediaStrategy which uses an existing IPC call.

Marking as failing in glib since there is a lack of mock rotation support.

Covered by added test.

* LayoutTests/fast/mediastream/image-capture-grabFrame-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-grabFrame.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::~ImageCapture):
(WebCore::videoFrameOrientation):
(WebCore::createImageBitmapViaDrawing):
(WebCore::createImageBitmapFromNativeImage):
(WebCore::createImageBitmap):
(WebCore::ImageCaptureVideoFrameObserver::create):
(WebCore::ImageCaptureVideoFrameObserver::~ImageCaptureVideoFrameObserver):
(WebCore::ImageCaptureVideoFrameObserver::add):
(WebCore::ImageCaptureVideoFrameObserver::stop):
(WebCore::ImageCaptureVideoFrameObserver::ImageCaptureVideoFrameObserver):
(WebCore::ImageCaptureVideoFrameObserver::processVideoFrame):
(WebCore::ImageCapture::grabFrame):
(WebCore::ImageCapture::stop):
(WebCore::ImageCapture::stopGrabFrameObserver):
* Source/WebCore/Modules/mediastream/ImageCapture.h:
* Source/WebCore/Modules/mediastream/ImageCapture.idl:
* Source/WebCore/platform/MediaStrategy.h:
(WebCore::MediaStrategy::nativeImageFromVideoFrame):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::orientationChanged):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::nativeImageFromVideoFrame):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:

Canonical link: <a href="https://commits.webkit.org/293243@main">https://commits.webkit.org/293243@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab5f181a6779fccb743215e63fe6eb5589063066

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17968 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103457 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48866 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100382 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18260 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26419 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/74854 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32016 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101341 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/13833 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88809 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55213 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13616 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6770 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48308 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6844 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105832 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25424 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83834 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25797 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85011 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83306 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21038 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27944 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5615 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19070 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25382 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25202 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28518 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26777 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson reviewed (approved):
Your commit message still says "WIP, needs more tests, especially for rotated camera frames"

youennf committed:
Add support for ImageCapture.grabFrame
rdar://148425176
https://bugs.webkit.org/show_bug.cgi?id=290916

Reviewed by Eric Carlson.

Add support of ImageCapture::grabFrame by doing the following:
- Introduce ImageCaptureVideoFrameObserver as a VideoFrameObserver on the ImageCapture track.
- Convert the VideoFrame into an ImageBitmap. We handle the rotation so that drawing the ImageBitmap is similar to rendering the track via a media element.
- The conversion requires conversion to RGB which is done in GPUProcess. For that purpose, we introduce a new MediaStrategy which uses an existing IPC call.

Marking as failing in glib since there is a lack of mock rotation support.

Covered by added test.

* LayoutTests/fast/mediastream/image-capture-grabFrame-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-grabFrame.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::~ImageCapture):
(WebCore::videoFrameOrientation):
(WebCore::createImageBitmapViaDrawing):
(WebCore::createImageBitmapFromNativeImage):
(WebCore::createImageBitmap):
(WebCore::ImageCaptureVideoFrameObserver::create):
(WebCore::ImageCaptureVideoFrameObserver::~ImageCaptureVideoFrameObserver):
(WebCore::ImageCaptureVideoFrameObserver::add):
(WebCore::ImageCaptureVideoFrameObserver::stop):
(WebCore::ImageCaptureVideoFrameObserver::ImageCaptureVideoFrameObserver):
(WebCore::ImageCaptureVideoFrameObserver::processVideoFrame):
(WebCore::ImageCapture::grabFrame):
(WebCore::ImageCapture::stop):
(WebCore::ImageCapture::stopGrabFrameObserver):
* Source/WebCore/Modules/mediastream/ImageCapture.h:
* Source/WebCore/Modules/mediastream/ImageCapture.idl:
* Source/WebCore/platform/MediaStrategy.h:
(WebCore::MediaStrategy::nativeImageFromVideoFrame):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::orientationChanged):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::nativeImageFromVideoFrame):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:

Canonical link: https://commits.webkit.org/293243@main
